store/postgres/src/sql/parser.rs

```diff
@@ -0,0 +1,191 @@
+use super::{constants::SQL_DIALECT, formatter::Formatter, validation::Validator};
+use crate::relational::{ColumnType, Layout};
+use anyhow::{anyhow, Ok, Result};
+use graph::components::store::BLOCK_NUMBER_MAX;
+use itertools::Itertools;
+use std::sync::Arc;
+
+pub fn generate_table_prelude_from_layout(layout: &Layout) -> String {
+    let schema = &layout.catalog.site.namespace;
+    let ctes = layout
+        .tables
+        .iter()
+        .filter(|(entity, _)| !entity.is_poi())
+        .map(|(_, table)| {
+            let table_name = table.name.as_str();
+
+            let (block_column, filter) = if !table.immutable {
+                (
+                    "block_range",
+                    Some(format!(" WHERE \"block_range\" @> {}", BLOCK_NUMBER_MAX)),
+                )
+            } else {
+                ("block$", None)
+            };
+
+            let columns = table
+                .columns
+                .iter()
+                .map(|col| {
+                    if !col.is_list() && col.column_type == ColumnType::Bytes {
+                        format!(
+                            r#"concat('0x', encode("{}", 'hex')) AS "{}""#,
+                            col.name.as_str(),
+                            col.name.as_str()
+                        )
+                    } else {
+                        format!(r#""{}""#, col.name.as_str())
+                    }
+                })
+                .chain(std::iter::once(format!(r#""{}""#, block_column)))
+                .collect::<Vec<_>>()
+                .join(", ");
+            format!(
+                "\"{table_name}\" AS (SELECT {columns} FROM \"{schema}\".\"{table_name}\"{})",
+                filter.unwrap_or_default()
+            )
+        })
+        .sorted()
+        .collect::<Vec<_>>()
+        .join(",\n");
+    format!("WITH {ctes}")
+}
```

Comment:
One of the requirements is that we didn't want to expose all the data and be able to "filter" what is exposed via SQL or not(i.e $block, etc). We had 3 possible solutions at the time:
- Use AST to keep track of all columns and reject via static analysis.
- Create Views
- Create CTEs

The AST route is by far the most efficient and most difficult to implement because we need to keep track of column names, possible rewrites and aliases, etc, and given the timeframe we decided to let it for future work. We had concerns with the view approach because we would need to handle "migrations" and update it for already deployed subgraphs.

The CTE route was the easiest and simplest way to create a "VIEW" per query and also rewrite/filter what we want.

Reply:
Yes, if the CTE is used only once, it gets inlined; if it is used more than once, it is materialized. Materializing these CTEs in general will be very expensive.

What I had in mind was rewriting something like `select * from tokens` to `select * from (select * from tokens where block_range @> $block) as tokens` with the hope that the Postgres optimizer can rewrite the query to move the where clause to the outer query if that seems to improve things, though I haven't tested that.

For the example queries you tried, how common are queries that use the same CTE more than once?

Reply:
Actually heavy as all CTEs are mapping to tables. On the other hand this is exactly conforms the non recursive and side-effect-free. So I expect optimizer to inline the CTE and merge filters. I haven't checked the execution plans but performance wise everything seems to be okey.